server/src/main/java/org/elasticsearch/indices/store/CompositeIndexFoldersDeletionListener.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.indices.store;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.plugins.IndexStorePlugin;
+
+import java.nio.file.Path;
+import java.util.List;
+
+public class CompositeIndexFoldersDeletionListener implements IndexStorePlugin.IndexFoldersDeletionListener {
+
+    private static final Logger logger = LogManager.getLogger(CompositeIndexFoldersDeletionListener.class);
+    private final List<IndexStorePlugin.IndexFoldersDeletionListener> listeners;
+
+    public CompositeIndexFoldersDeletionListener(List<IndexStorePlugin.IndexFoldersDeletionListener> listeners) {
+        for (IndexStorePlugin.IndexFoldersDeletionListener listener : listeners) {
+            if (listener == null) {
+                throw new IllegalArgumentException("listeners must be non-null");
+            }
+        }
+        this.listeners = List.copyOf(listeners);
+    }
+
+    @Override
+    public void beforeIndexFoldersDeleted(Index index, IndexSettings indexSettings, List<Path> indexPaths) {
+        for (IndexStorePlugin.IndexFoldersDeletionListener listener : listeners) {
+            try {
+                listener.beforeIndexFoldersDeleted(index, indexSettings, indexPaths);
+            } catch (Exception e) {
```

Comment:
This is a bit trappy IMO, do we permit these methods to throw exceptions or not? It looks like callers don't expect exceptions, but implementors expect to be able to throw them, so we must be sure always to use a `CompositeIndexFoldersDeletionListener` in the middle to swallow any exceptions.

(I know we have this pattern in other `Composite...` places but that doesn't mean I agree with it)

Could we not swallow exceptions here (and maybe assert no exception is thrown). Alternatively could we make the field in `IndicesService` into a `CompositeIndexFoldersDeletionListener` rather than a `IndexStorePlugin.IndexFoldersDeletionListener` to spell out that the exception-swallowing behaviour is required.

Reply:
I agree this is trappy. Looking at the places where the `IndexFoldersDeletionListener` are expected to be called I think we can just not swallowing them in `CompositeIndexFoldersDeletionListener` (I pushed d6a6e2dfc48c43287137501afbf8dfc8502b7b18) and allow listeners to throw exceptions.